### DISCUSSION TITLE
Is it possible to use another hosting provider than Vercel with incremental builds?

### DISCUSSION THREAD

**Discussion by jordyvandomselaar** [vercel/next.js#13670] [Help]
The company I work for is looking into AWS and would prefer to manager everything through AWS. Is it possible to use incremental builds/static exports with AWS?

**Comment by balazsorban44** [comment]
https://github.com/danielcondemarin/serverless-next.js

This might help, not sure about the incremental bit. Right now it looks like Next.js's success heavily relies on tight integration with the Vercel platform, and there is no official support for other hosting solutions with all the same benefits.

**Reply by timneutkens** [reply]
`next start` supports incremental generation and regeneration

**Reply by Timer** [reply]
> Right now it looks like Next.js's success heavily relies on tight integration with the Vercel platform, and there is no official support for other hosting solutions with all the same benefits.

Just to reiterate the above point from @timneutkens, the above quote is completely untrue.

All Next.js features work with any hosting provider that supports `next start`. You can use incremental builds with AWS and a CDN.

**Reply by balazsorban44** [reply]
Yes, I am aware, and very grateful for that! We are using it as well, hosting our nextjs app on Azure with a custom server.

It has been an ongoing wish of mine to get to model/mirror the amazing Vercel platform for those of us who are locked in on other services like AWS and Azure, but would like to squeeze out as much of the framework as Vercel is capable of. I am totally aware why Vercel/Next.js is coupled like this, and have nothing against this individually. But as my company is not willing to turn away from Azure (probably the same for other enterprises and their own sevcies), I find it difficult to persuade my team about the advantages of being able to serve static files from a CDN where it is possible, reducing cost and complexity.

**Reply by balazsorban44** [reply]
I am sorry if this came out negatively. I try to advocate for using Next.js whenever I can, because I love it. It is more of a frustration in my professional/work life that we cannot utilise Vercel than anything against the framework itself. I am ever grateful for that we can still use everything with the overhead of having an extra Node.js server working in the background.

**Reply by balazsorban44** [reply]
You probably understand that questions like these are raised, because there is currently no official "easy" support for other platforms with the same benefits. Yes, there are ways, but all of them require custom/unofficial solutions to be able to deploy Next.js the same way as on Vercel. The above link is an example. The disadvantage of solutions like these that they are always a bit behind of what Next.js+Vercel can offer.


Trying not to make it off topic too much, 

*yes* as @timneutkens and @Timer suggests, *you can use any runtime that supports Node.js, and serve everything with `next start`*

But looking at the docs a bit, you find that using other platforms than Vercel:

> Next.js can be deployed to any hosting provider that supports Node.js. This is the approach you should take if you’re using a custom server.

When generally the advice is to try to avoid a custom server IF possible.


(@Timer @timneutkens if you have a bit of spare time and/or willing to discuss this in private, I would be ever grateful for some conversation around this on Twitter or in email.)

**Reply by jordyvandomselaar** [reply]
> `next start` supports incremental generation and regeneration

That's really cool =), when using `next start`, I assume it exports static files as it goes then? So then we would need a watcher to check if there are new files to send them through a CDN?

**Reply by balazsorban44** [reply]
I had a chat with @Timer, and he convinced me that **`next start` is a great option** if choosing other platforms and my previous answers were based on my bad understanding around the docs.

Yes, it generates static files in the .next folder. After talking with @Timer, as I understood, ALL your requests to Next.js can be sent through a CDN.

**Reply by jordyvandomselaar** [reply]
But you would need `next start` to be running which you can't do on a CDN? You would need to send the contents of `.next` to a CDN?

**Reply by timneutkens** [reply]
You need an enterprise-grade CDN that supports caching dynamic responses (which is expected if you're going to host any app yourself). Vercel provides an enterprise-grade CDN to you out of the box for free and integrates with Next.js automatically when you deploy on Vercel.

**Reply by jordyvandomselaar** [reply]
Thanks for the explanation! I'm a happy user of Vercel myself and I'm just considering our options for when I pitch the idea for projects =)

**Comment by JonCognioDigital** [comment]
I'm not sure I understand the bit about "Next Start", AFAIK that's just a CLI command used to start Next?

I have a NextJS site hosted on Azure and have the ability to call a webhook from my CMS when content is updated but how could that webhook be set up and how could it trigger a re-build on an already deployed site?

**Reply by balazsorban44** [reply]
This is not yet implemented, but it may be available in the future: https://github.com/vercel/next.js/discussions/11552#discussioncomment-2655

What you have for now is a time-based revalidation.

Meaning, if you return `unstable_revalidate: 1` (the number being the time in seconds after which you want to revalidate) from `getStaticProps`, the page will be regenerated IN THE BACKGROUND. So the user will always be served with a static file.


Just an idea:
You could maybe trigger an update anyway, if your CMS changes, you can then send a crawler to that page of your site, so the "next real user" will have the same updated version as in the CMS. This is not tested, and I just assume this would work. @Timer what do you think?

**Reply by JonCognioDigital** [reply]
I like the crawler idea. I assume that once the revalidate time expires (which could be Googlebot) they will still get a statically served page, it just might not be up to date?

**Reply by revskill10** [reply]
@balazsorban44  Good idea, but i think it currently only solve the `fallback: true` part for newly created content, not the `revalidation` part. We still have to wait till `event -based revalidation trigger ` implemented then a crawler will achive our result :)

So, with that idea, i think we need a special route to trigger validate the pages by crawling the path.

Something like this might work:
```js
export const getStaticPaths() {
  const paths = ...
  const revalidatedPaths = ... // same format as paths
  return {
    paths,
    revalidate: revalidatePaths
  }
}
```
So only request to `revalidatePaths` will trigger the regeneration when in need.
But those `revalidatePaths` need to keep in private with a private header key,... so that we could setup an event trigger from CMS to post to those `revalidatePaths`

@timneutkens How do you think about this idea ?

**Reply by balazsorban44** [reply]
I would assume the event comes from your CMS, not from Next. So my idea:

1. change in CMS
2. a crawler picks up the event, "page x has been updated in CMS"
3. crawler visits your page x on Next.js
4. revalidation in the background
5. user gets fresh page

**Reply by balazsorban44** [reply]
This might not be the right thread to discuss this though, there is #11552. I just had a quick idea for @jon64digital's question.

**Comment by usagi-f** [comment]
I want to make the Incremental Static Regeneratoin work properly with AWS solution, but I'm concerned about the relationship between the number of seconds set by revalidate and the TTL setting of CDN (actually Cloudfront).

If CDN has a long TTL more than revalidate value, even if regenerated, then I think the cache(old html) will be prioritized.

However, If set the short revalidate value(e.g. 1sec) and also make TTL is same, the original merit of the CDN will be lost.

My understanding has any structual mistakes?
